### ISSUE TITLE
fatal: Authentication failed

### ISSUE THREAD

**Issue by wangqinhu** [yuki-kimoto/gitprep#38]
Hi Yuki,

GitPrep is really great project. I setup it on Mac OS X (10.9.1, local) and Linux (Cent OS 6.4, server), when I use git push, both local and server return the following message:

Username for 'http://xxx.xxx.xxx.xxx': username
Password for 'http://username@xxx.xxx.xxx.xxx':
fatal: Authentication failed for 'http://xxx.xxx.xxx.xxx/gitprep.cgi/username/repox.git/'

I am sure I have the right username and password, and I also can access the server via ssh key, would you please help me to figure out the problem?

BTW, there is no problem when setup under a particular user home directory.

Thanks,

Qinhu.

**Comment by cyberlink1** [comment]
I am having the same issue and have been digging into it.

Log:
[Mon Jan 18 15:58:40 2016] [info] User:, Password:
[Mon Jan 18 15:58:40 2016] [warn] Authentication: User name is empty

----------------------------------------
Server version: Apache/2.4.17 
Running getprep as a CGI
git version 2.4.3
perl v5.20.3
Fedora Core 22 - x86_64
----------------------------------------
/etc/httpd/conf.d/gitprep.conf

<Directory /var/www/html/gitprep>
    Options +ExecCGI
    AddHandler cgi-script .cgi
</Directory>
-----------------------------------------

I have noticed that when I try to check out a private repo or attempt to check in any repo via http, It prompts me for an openssh login.  If I put http://username:password@host/gitprep/name/repo it 
It immediately fails with "fatal: Authentication failed for"

It is almost as if it is not getting to the http auth and is jumping directly to an ssh auth.

**Comment by yuki-kimoto** [comment]
I think this bug is client side. I don't solve this problem from now.

Please try http://username@host/gitprep/name/repo

What result is returned?

**Comment by yuki-kimoto** [comment]
and please tell me http URL.

**Comment by cyberlink1** [comment]
yuki san,

Using http://username@host/gitprep/name/repo
causes it to open a request on my system for the ssh password. It does not matter what is entered it returns "fatal: Authentication failed for <username>" (My username is displayed where I put <username>)

**Comment by ckozler** [comment]
This is happening for me too. 

Client side

:~/Code > git --version
git version 2.4.3

Server side

:/var/www/html/git/log]# git --version
git version 1.7.1

**Comment by yuki-kimoto** [comment]
I haven't solved  this problem yet. Please tell me information.

1. Git Version (Client and Server)
2. OS and OS version (Client and Server)
3. curl version (Client)

**Comment by cyberlink1** [comment]
Git Client:  git version 2.4.3
Git Server: git version 2.4.3
Os Version Server: Fedora release 22 (Twenty Two)
Os Version Client: Fedora release 22 (Twenty Two)
curl version Client: curl 7.40.0 
curl version Server: curl 7.40.0

**Comment by yuki-kimoto** [comment]
I want to copy the following code and paste it in GitPrep.pm

    # GitPrep.pm line 249
    if ($service eq 'git-receive-pack' || $private) {
      
      # Pelase add the following code
      my $req_userinfo = $self->req->url->to_abs->userinfo;
      my $req_url_abs = $self->req->url->to_abs;
      my $req_url = $self->req->url;
      $self->app->log->info($req_userinfo);
      $self->app->log->info($req_url_abs);
      $self->app->log->info($req_url);
      # End
      
      $self->basic_auth("Git Area", sub {
        my ($auth_user, $auth_password) = @_;
        
        if (!defined $auth_user || !length $auth_user) {
          $self->app->log->warn("Authentication: User name is empty");
        }
        
        $auth_user = '' unless defined $auth_user;
        $auth_password = '' unless defined $auth_password;
        
        my $is_valid =
          ($user eq $auth_user || $api->is_collaborator($user, $project, $auth_user))
          && $api->check_user_and_password($auth_user, $auth_password);
        
        return $is_valid;
      });
    }

**Comment by yuki-kimoto** [comment]
@wangqinhu 
@adamkruszewski 
@knutov 
@cyberlink1 
@ckozler 

I released GitPrep 2.0. I want to know GitPrep 2.0 have this bug yet.
If you have some time, please try GitPrep 2.0.

**Comment by paulmaxwell** [comment]
Hi Yuki.

I was testing GitPrep 2.2.1 and I'm getting same authentication error. When I'm trying to do git push (via http), I'm getting this error

```
error: RPC failed; curl 56 Illegal or missing hexadecimal sequence in chunked-encoding
```

The log says:

```
[Mon Jun 27 08:11:54 2016] [warn] Authentication: User name is empty
```

Software versions:

 - server

  - git 2.1.4
  - perl 5.20.2
  - os: debian 8, linux 3.18.25-gentoo-r1 (in lxc container)

 - client

  - git 2.9.0
  - curl 7.49.1
  - os: archlinux, linux 4.3.3

I'm not sure, but I suppose the issue is on **Mojolicious** side, seems like it can't decode username and password from **Authentication** header for some reason. I'm sorry but I had no chance to check it.

I hope this will help.

PS You're doing great work. I like GitPrep :)

**Comment by thigas88** [comment]
I'm having the same problem.

**Comment by darkz1984** [comment]
I have the same problem too,and some repo have the authencation problem,and other is ok...

**Comment by darkz1984** [comment]
I found I have no privileges of other's project, add me to the Collaborators of the project ,and is ok，and I upgrade gitprep recently,so I think maybe it has some problem of the upgrade...

**Comment by yuki-kimoto** [comment]
@paulmaxwell @thigas88

Thanks for bug reporting.

I don't yet understand why basic authentication fail in only some CGI environments.

**Comment by wangqinhu** [comment]
Hi Yuki,

I haven't use gitprep for more than one year, and I tried your latest gitprep (v2.3) just now. Fortunately, I found the UI and functions are greatly improved. More importantly, the bug described in this issue is nolonger appear in my new environment.

Here are the details of my environment:

```
# Server and Client
- OS: OS X (10.11.4)
- git: 2.7.4
- perl: 5.18.2
- curl: 7.43.0
```


Regards,

Qinhu

**Comment by yuki-kimoto** [comment]
@wangqinhu 

thanks for your report!

**Comment by necrophcodr** [comment]
The problem happens on Apache2.2 and Apache2.4 on Ubuntu 12.04 and Ubuntu 14.04 and Ubuntu 16.04 with virtualmin installed and gitprep set up as CGI.

Solution is to change .htaccess as follows:

    deny from all
    
    <Files ~ "gitprep\.cgi">
    allow from all
    </Files>
    
    SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1
    
    DirectoryIndex index.cgi

After adding the

    SetEnvIf Authorization "(.*)" HTTP_AUTHORIZATION=$1

line, authorization headers are properly passed to the cgi script.

**Comment by yuki-kimoto** [comment]
Thank you! I will add that line.

**Comment by toshi-ara** [comment]
I send a Pull request.

**Comment by Zero3K** [comment]
Any news regarding this issue?

**Comment by Jade-Wang** [comment]
Did you ever setting up 2FA authentication on github?
If so, can refer to https://medium.com/@ginnyfahs/github-error-authentication-failed-from-command-line-3a545bfd0ca8 to use personal access token instead of pasword.

**Comment by pkj-01** [comment]
I came to the same question, do you know anything about how to solve this situation? ...

**Comment by yuki-kimoto** [comment]
@pkj-01 

I am sorry. The problem is probably one that is difficult to find a solution for and I do not have time to investigate at this time.

**Comment by pkj-01** [comment]
I understand. I will continue to investigate in the same way and if I find a solution, I will communicate it. Thank you ...

**Comment by gstrauss** [comment]
FYI: Apache2 and `Authorization` not sent to CGI:
https://linux.m2osw.com/http-authentication-header-missing-not-sent-my-cgi-script-why

**Comment by necrophcodr** [comment]
> FYI: Apache2 and `HTTP_AUTHORIZATION` not sent to CGI: https://linux.m2osw.com/http-authentication-header-missing-not-sent-my-cgi-script-why

This is what I wrote to do as well.

**Comment by gstrauss** [comment]
@necrophcodr  Yes.

The link I gave says that the Apache2 behavior is intentional and *is not* an accidental bug in Apache2.
To get `HTTP_AUTHORIZATION` set in the CGI environment created by Apache2, you have to explicitly add one of the workarounds (e.g. posted by you or @yuki-kimoto to the Apache httpd.conf or .htaccess (if enabled).

(If https://github.com/toshi-ara/gitprep/commit/c11e08673a4bfa567309831c579a738bb9628bb6 fixes this, then should this issue be closed?)

**Comment by monnerat** [comment]
A fix for this bug has been committed on Nov 12, 2018 and is included in the v2.6.2 release: https://github.com/yuki-kimoto/gitprep/commit/c11e08673a4bfa567309831c579a738bb9628bb6
If the problem persists, please reopen this issue.
